source/BankPanel.cpp

```diff
@@ -108,9 +121,10 @@ void BankPanel::Draw()
 	static const string prefix[2] = {"salary: ", "tribute: "};
 	for(int i = 0; i < 2; ++i)
 	{
-		auto it = player.Conditions().lower_bound(prefix[i]);
-		for( ; it != player.Conditions().end() && !it->first.compare(0, prefix[i].length(), prefix[i]); ++it)
-			income[i] += it->second;
+		CollapseConditions(player.Conditions(), prefix[i],
```

Comment:
If generalizing this sum, I would expect us to use the standard library rather than reinvent the wheel. i.e. use `find_if`/`find_if_not` to determine the end of the range, and then pass the iterators to `std::accumulate`.

Reply:
The goal of this PR was not to make the iterators nicer, it was to check if we can get rid of the iterators completely for code that uses player-conditions.

Do we still want player.Conditions() to be iteratable after #5017? And would the iterator keys still need to be string (and sorted by string) or would a more complex type like pair<list-identifier, string> also be okay?

Reply:
I would indeed expect that an iteration interface is exposed. Whatever solution we go with should be able to provide a `begin` and an `end` ForwardIterator; they don't have to point to the same container (but obviously would then need to implement the increment operator correctly).

Reply:
If we implement this iterator, should it then be an iterator over just the manually set conditions, or should it also iterate over all the automatically on-demand generated conditions?
I feel that generating conditions on-demand to just iterate over them and then discard the value if the key doesn't match is not a good use of CPU power.

Reply:
That's a good point. Can the key and value be generated separately? That is, will we be able to identify the available condition keys without computing their corresponding values?

Dereferencing the iterator would then yield a function that returns the condition value, rather than the condition value itself.

Reply:
Yes, I think that would technically be possible, but what does it then mean to iterate over on-demand conditions?

Suppose that we at some point support get on-demand conditions like `named ship: [Ship Name]: Outfit: Energy Blaster` which tells us how many Blasters are installed on a ship with a certain name. Or even conditions like `tagged ships: [tag id]: Outfit: Energy Blaster` which tell us how many Blasters there are installed in total on a set of ships that share the same tag.

If we would want to iterate over this type of data in the C++ code, then I would expect us to iterate directly over the ships and not to iterate over the condition sets.

So far I only found 2 locations where there was an iteration over the conditions set:
- When the conditions are saved to a savegame (which only requires the manual conditions, not the on-demand ones)
- When the salaries and tributes are retrieved for display (which currently uses manual conditions, but might move to their own interatable class in the future)

When would we want to iterate over all possible (manual and on-demand) conditions?